libraries/common-arm/src/sd_manager.rs

```diff
@@ -52,85 +56,126 @@ where
+            Ok(file) => Some(file),
             Err(_) => {
-                panic!("Cannot create file.");
+                error!("Cannot create file.");
+                None
             }
         };
 
         SdManager {
-            sd_controller: sd_cont,
+            sd_controller,
             volume,
             root_directory,
-            file: Some(file),
+            file,
         }
     }
     pub fn write(
         &mut self,
         file: &mut sd::File,
         buffer: &[u8],
     ) -> Result<usize, sd::Error<sd::SdMmcError>> {
-        self.sd_controller.write(&mut self.volume, file, buffer)
+        if !self.is_mounted() {
+            return Err(sd::Error::DeviceError(sd::SdMmcError::CardNotFound));
+        }
+        self.sd_controller.write(self.volume.as_mut().unwrap(), file, buffer)
     }
     pub fn write_str(
         &mut self,
         file: &mut sd::File,
         msg: &str,
     ) -> Result<usize, sd::Error<sd::SdMmcError>> {
+        if !self.is_mounted() {
+            return Err(sd::Error::DeviceError(sd::SdMmcError::CardNotFound));
+        }
         let buffer: &[u8] = msg.as_bytes();
-        self.sd_controller.write(&mut self.volume, file, buffer)
+        self.sd_controller.write(self.volume.as_mut().unwrap(), file, buffer)
     }
     pub fn open_file(&mut self, file_name: &str) -> Result<sd::File, sd::Error<sd::SdMmcError>> {
+        if !self.is_mounted() {
+            return Err(sd::Error::DeviceError(sd::SdMmcError::CardNotFound));
+        }
         self.sd_controller.open_file_in_dir(
-            &mut self.volume,
-            &self.root_directory,
+            self.volume.as_mut().unwrap(),
+            self.root_directory.as_ref().unwrap(),
             file_name,
             sd::Mode::ReadWriteCreateOrTruncate,
         )
     }
     pub fn close_current_file(&mut self) -> Result<(), sd::Error<sd::SdMmcError>> {
+        if !self.is_mounted() {
+            return Err(sd::Error::DeviceError(sd::SdMmcError::CardNotFound));
+        }
         if let Some(file) = self.file.take() {
             return self.close_file(file);
         }
         Ok(())
     }
     pub fn close_file(&mut self, file: sd::File) -> Result<(), sd::Error<sd::SdMmcError>> {
-        self.sd_controller.close_file(&self.volume, file)
+        if !self.is_mounted() {
+            return Err(sd::Error::DeviceError(sd::SdMmcError::CardNotFound));
+        }
+        self.sd_controller.close_file(self.volume.as_ref().unwrap(), file)
     }
-    pub fn close(mut self) {
-        self.sd_controller
-            .close_dir(&self.volume, self.root_directory);
+    pub fn close(&mut self) -> Result<(), sd::Error<sd::SdMmcError>> {
+        if !self.is_mounted() {
+            return Err(sd::Error::DeviceError(sd::SdMmcError::CardNotFound));
+        }
+        self.sd_controller.close_dir(
+            self.volume.as_ref().unwrap(), 
+            self.root_directory.as_ref().unwrap()
```

Comment:
@NoahSprenger So `sd_controller.close_dir` for some reason doesn't take a `&Directory dir` as the second param but `Directory dir` unwrapping without `as_ref` throws a different error

```
cannot move out of `self.root_directory` which is behind a mutable reference
help: consider calling `.as_ref()` or `.as_mut()` to borrow the type's contents
```

Reply:
If I don't make the `root_directory` mutable then I can't do the trick I did up there where I initialize the `root_directory` after I checked to see `volume` is already setup. SInce it's assigning None first then the actual `Directory` once the check is successful